pkg/aws/eni/node.go

```diff
@@ -446,6 +477,73 @@ func (n *Node) determineAllocationAction() (*allocatableResources, error) {
 		limits:     limits,
 		totalENIs:  len(n.enis),
 	}
+
+	// Validate that the node still requires addresses to be released, the
+	// request may have been resolved in the meantime.
+	if option.Config.AwsReleaseExcessIps && n.stats.excessIPs > 0 {
+		for key, e := range n.enis {
+			scopedLog.WithFields(logrus.Fields{
+				fieldEniID:     e.ID,
+				"needIndex":    n.resource.Spec.ENI.FirstInterfaceIndex,
+				"index":        e.Number,
+				"addressLimit": a.limits.IPv4,
+				"numAddresses": len(e.Addresses),
+			}).Debug("Considering ENI for IP release")
+
+			if e.Number < n.resource.Spec.ENI.FirstInterfaceIndex {
+				continue
+			}
+
+			// Find free IP number on this ENI
+			ipsOnENI := n.resource.Status.ENI.ENIs[e.ID].Addresses
+			freeIpsOnENI := []string{}
+			for _, ip := range ipsOnENI {
+				_, ipUsed := n.resource.Status.IPAM.Used[ip]
+				// exclude primary IPs
+				if !ipUsed && ip != e.IP {
+					freeIpsOnENI = append(freeIpsOnENI, ip)
+				}
+			}
+			freeOnENICount := len(freeIpsOnENI)
+
+			if freeOnENICount <= 0 {
+				continue
+			}
+
+			scopedLog.WithFields(logrus.Fields{
+				fieldEniID:       e.ID,
+				"excessIPs":      n.stats.excessIPs,
+				"freeOnENICount": freeOnENICount,
+			}).Debug("ENI has IPs free")
+			maxReleaseOnENI := math.IntMin(freeOnENICount, n.stats.excessIPs)
+
+			if a.ipsToReleaseOnENI == nil || maxReleaseOnENI > len(a.ipsToReleaseOnENI) {
+				a.eni = key
+				a.subnet = &types.Subnet{ID: e.Subnet.ID}
+				a.ipsToReleaseOnENI = freeIpsOnENI[:maxReleaseOnENI]
+			}
+		}
+		scopedLog = scopedLog.WithFields(logrus.Fields{
+			"available":      n.stats.availableIPs,
+			"used":           n.stats.usedIPs,
+			"excess":         n.stats.excessIPs,
+			"releasing":      a.ipsToReleaseOnENI,
+			"selectedENI":    n.enis[a.eni],
+			"selectedSubnet": a.subnet.ID,
+		})
+
+		scopedLog.Info("Releasing excess IPs from node")
```

Comment:
Make this log message dependant on `a.ipsToReleaseOnENI != nil` as otherwise we might indicate that we are releasing IPs when we are in fact not.

Reply:
done